fix: validate UTF-8 at C++ -> Lean boundary

digama0 opened the pull request:
Continuation of #3958. To ensure that lean code is able to uphold the invariant that `String`s are valid UTF-8 (which is assumed by the lean model), we have to make sure that no lean objects are created with invalid UTF-8. #3958 covers the case of lean code creating strings via `fromUTF8Unchecked`, but there are still many cases where C++ code constructs strings from a `const char *` or `std::string` with unclear UTF-8 status.

To address this and minimize accidental missed validation, the `(lean_)mk_string` function is modified to validate UTF-8. The original function is renamed to `mk_string_unchecked`, with several other variants depending on whether we know the string is UTF-8 or ASCII and whether we have the length and/or utf8 char count on hand. I reviewed every function which leads to `mk_string` or its variants in the C code, and used the appropriate validation function, defaulting to `mk_string` if the provenance is unclear.

This PR adds no new error handling paths, meaning that incorrect UTF-8 will still produce incorrect results in e.g. IO functions, they are just not causing unsound behavior anymore. A subsequent PR will handle adding better error reporting for bad UTF-8.

digama0 commented:
!bench

digama0 commented:
(disabled auto-merge to run a benchmark first)

digama0 commented:
The affected tests are regression tests from #301, #1707, #1690, all of which appear to have been strange behaviors of lean caused by invalid UTF-8. The new behavior is for the C code to throw an unspecific error before even calling lean code; this is inevitable because the lean parser/elaborator assumes it can store the file in a `String` so we can't really use most of lean's error handling.

nomeata commented on src/include/lean/lean.h:
```diff
@@ -2064,11 +2066,11 @@ static inline uint8_t lean_version_get_is_release(lean_obj_arg _unit) {
 
 static inline lean_obj_res lean_version_get_special_desc(lean_obj_arg _unit) {
-    return lean_mk_string(LEAN_SPECIAL_VERSION_DESC);
+    return lean_mk_ascii_string(LEAN_SPECIAL_VERSION_DESC);
```
I’d probably lean towards using `mk_utf8_string` here. It’s probably not performance critical, and maybe some German developer with a ü in their name builds lean from a branch with `v0.7.8-Günnar`….
```suggestion
    return lean_mk_utf8_string(LEAN_SPECIAL_VERSION_DESC);
```

nomeata commented on src/include/lean/lean.h:
```diff
@@ -2064,11 +2066,11 @@ static inline uint8_t lean_version_get_is_release(lean_obj_arg _unit) {
 
 static inline lean_obj_res lean_system_platform_target(lean_obj_arg _unit) {
-    return lean_mk_string(LEAN_PLATFORM_TARGET);
+    return lean_mk_ascii_string(LEAN_PLATFORM_TARGET);
```
```suggestion
    return lean_mk_utf8_string(LEAN_PLATFORM_TARGET);
```
Unlikely to matter, but better safe than sorry.

digama0 replied:
well, the "better safe than sorry" option is `lean_mk_string`. If we have any reason to believe that the data isn't necessarily in the subset it appears to be in, we should just validate it.

nomeata replied:
Indeed, after I wrote this and was away from the computer I paused wondered if my suggestion is bogus.

It’s not entirely bogus, though: If the there is utf8 in the TARGET, then `mk_ascii_string` will produce an invalid length. So if `LEAN_SPECIAL_VERSION_DESC` etc. happens to be utf8, then `mk_utf8_string` is better than `mk_ascii_string`, but not as safe as a validating function.

I agree that it’s worth changing the `lean_mk_{utf8,utf8}_string` function names to include `unchecked to avoid future bugs.

Kha replied:
Agreed on better names

nomeata commented on src/runtime/platform.cpp:
```diff
@@ -40,7 +40,7 @@ extern "C" LEAN_EXPORT uint8 lean_system_platform_emscripten(obj_arg) {
 }
 
-extern "C" object * lean_get_githash(obj_arg) { return lean_mk_string(LEAN_GITHASH); }
+extern "C" object * lean_get_githash(obj_arg) { return lean_mk_ascii_string(LEAN_GITHASH); }
```
Also unlikely, but since the source is out of our (local) control, better safe than sorry.
```suggestion
extern "C" object * lean_get_githash(obj_arg) { return lean_mk_utf8_string(LEAN_GITHASH); }
```

Kha commented on src/include/lean/lean.h:
```diff
@@ -991,6 +991,8 @@ static inline size_t lean_string_byte_size(lean_object * o) { return sizeof(lean
 /* instance : inhabited char := ⟨'A'⟩ */
 static inline uint32_t lean_char_default_value() { return 'A'; }
 LEAN_EXPORT lean_obj_res lean_mk_string_from_bytes(char const * s, size_t sz);
+LEAN_EXPORT lean_obj_res lean_mk_ascii_string(char const * s);
```
Are we using this in any performance-critical spot?

nomeata replied:
I introduced that function for Nat.repr

Kha commented on src/runtime/object.cpp:
```diff
@@ -1611,9 +1611,23 @@ extern "C" LEAN_EXPORT object * lean_mk_string_from_bytes(char const * s, size_t
 
 extern "C" LEAN_EXPORT object * lean_mk_string(char const * s) {
+    size_t len = strlen(s);
+    if (validate_utf8((const uint8_t *)s, len)) {
```
Why is this a separate step, can't it be combined with `utf8_strlen`? Do we even need the mk_string/mk_utf8_string distinction in that case?

digama0 replied:
We can do `utf8_strlen` at the same time as validation, but I think there are still use cases for "I know this is UTF8", in particular when doing the check beforehand so as to give a custom error message. Also, from the looks of it `utf8_strlen` will be optimized to a faster function than `validate_utf8`, so there is likely some performance advantage to using only `utf8_strlen` when there are reasons to believe the input is already utf8. This happens in `String.set` for example (although actually in that example we can do even better since we know what the length is already).

Kha commented on src/runtime/object.cpp:
```diff
@@ -1611,9 +1611,23 @@ extern "C" LEAN_EXPORT object * lean_mk_string_from_bytes(char const * s, size_t
+    if (validate_utf8((const uint8_t *)s, len)) {
+        return lean_mk_string_from_bytes(s, len);
+    } else {
+        return lean_mk_string_core("", 0, 0);
```
Is this really a sensible fallback? I matches the panic value but there is no message. Shouldn't this raise an exception if done as part of I/O?

digama0 replied:
Not really. I was planning to put in a follow-up PR which does something like rust's `String::from_utf8_lossy` with replacement characters. In most cases it's not really a good idea to rely on this for error handling, and instead code should call `validate` separately and throw IO errors etc as appropriate.

Kha replied:
I think it would be good to elaborate on this plan before going forward. It would also be helpful to mention such relevant plans in the PR description :) .

If we want to get to the point where no-one, especially not Lean APIs, depends on this branch, should it become a hard assertion?

digama0 replied:
I have a revision in the works which addresses many of the review comments. I'll post that and then we can talk about how we want to proceed here.

digama0 commented:
I pushed a new version of the PR. Summary of changes:

* `lean_mk_string_unchecked(const char * s, size_t sz, size_t utf8_chars)` was previously `lean_mk_string_core`
  * Use this when we have both the length and utf8_length values available and also know the input is valid UTF-8.
* `lean_mk_string_from_bytes_unchecked(const char * s, size_t sz)` was `lean_mk_string_from_bytes`
  * Use this when we have the length and know the input is valid UTF-8, but need to compute the utf8_chars.
* `lean_mk_string_from_bytes(const char * s, size_t sz)` does validation now, and if validation fails it converts invalid UTF-8 sequences to `U+FFFD` instead of just returning the empty string.
  * Use this when we have a (char*, len) pair but do not know it encodes valid UTF-8
* `lean_mk_string(const char * s)` is a wrapper around `lean_mk_string_from_bytes`
  * Use this when we have char*, not necessarily UTF-8 and need `strlen` to get the length
* `lean_string_from_utf8` was renamed to `lean_string_from_utf8_unchecked` 
  * This is the implementation of `String.fromUTF8`, you should probably not call it directly. I added `_unchecked` to the name because the C version doesn't validate, even though the lean version takes a validity proof witness as an argument.
* `lean_mk_ascii_string_unchecked(char const * s)` was `lean_mk_ascii_string`
  * Use this when you know the string is valid ASCII but need `strlen` to get the length.
* `lean::mk_ascii_string_unchecked(std::string const & s)` was `lean::mk_ascii_string`
  * Use this when you know the string is valid ASCII and have a `std::string` (which counts the length).

Other (non-API) changes:

* The code generator produces calls to `lean_mk_string_unchecked` now instead of `lean_mk_string_from_bytes`, since we don't want to validate string literals, plus we can calculate the utf8_chars in advance.
* ` bool validate_utf8(uint8_t const * str, size_t size, size_t & pos, size_t & i)` also returns the byte position of the first bad character, which can be used for better error reporting in addition to its use here as a fast path for the replacement character algorithm.
* I removed the special error handling for invalid UTF-8 in lean files. These cases, in addition to the IO uses of `mk_string`, are handled by the replacement algorithm, meaning that lean will complain in the bad UTF-8 tests with a "more normal" error saying that `U+FFFD` is not a token; if you stick it inside a comment or guillemets it will be accepted like any other unicode character. This is not necessarily a good choice, but at least it's a less damaging default behavior than erasing the whole string. You can however do separate validation, by calling `validate_utf8` and `lean_mk_string_unchecked` instead of `lean_mk_string`.

nomeata commented:
Sounds like a nice and thorough cleanup, thanks a lot!

I'd lean towards rejecting invalid utf8 in lean files (not replacing, not using the empty string). Who knows what kind of surprises will occur when people accidentally have invalid utf8 and lean does something, but not what they expect to.

Kha commented:
I'm happy as well modulo below -

> I'd lean towards rejecting invalid utf8 in lean files (not replacing, not using the empty string)

Agreed, and more generally I/O functions should at least by default throw an error when decoding fails. I'm pretty sure this is the default in all the big stdlibs out there as well, no?

digama0 commented:
I agree, but can we agree that this is an improvement on the status quo and do that in a follow up PR? Handling UTF-8 errors properly means having new kinds of error reporting functions we don't have currently, as well as touching a bunch more code.

Kha commented:
Sure, I just want to make sure it is documented and discussed that this is a partial solution in that way. That is very relevant context for reviewing.

nomeata commented:
Sounds like as soon as the PR description reflects the current state I can press the button?

digama0 commented:
PR description updated

nomeata commented:
Linux-debug fails with 
```
1658/1862 Test #1661: leanruntest_subset.lean ...................................***Failed    1.07 sec
../../common.sh: line 42: 48489 Aborted                 (core dumped) LEAN_BACKTRACE=0 "$@" 2>&1
     48490 Done                    | perl -pe 's/(\?(\w|_\w+))\.[0-9]+/\1/g' > "$f.produced.out"
Unexpected return code 134 executing 'lean -Dlinter.all=false subset.lean'; expected 0. Output:
LEAN ASSERTION VIOLATION
File: /home/runner/work/lean4/lean4/build/stage1/include/lean/lean.h
Line: 473
lean_is_string(o)
(C)ontinue, (A)bort/exit, (S)top/trap
```

Kha commented:
@digama0 Do you have time to look into the assertion failure?

digama0 commented:
Sorry for the delay, the assertion failure should be fixed now.
